icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -530,8 +548,116 @@ class Token {
         return fMatch - kSimpleUnitOffset;
     }
 
+    // TODO: Consider split function as a utility function.
+    // Parse the given string to a unsigned long value.
+    // If the value is not positive integer, it will return `kUnitIdentifierSyntaxError`.
+    uint64_t parseStrigToLong(const StringPiece str, UErrorCode &status) {
+        uint64_t result = 0;
+        uint64_t max_int64 = ~((int64_t)1 << 63); // the maximum number of int64_t
+
+        // Check for empty string
+        if (str.empty()) {
+            status = kUnitIdentifierSyntaxError;
+            return result;
+        }
+
+        int32_t exponent = 0;
+        int32_t exponentIndex = str.length();
+
+        // Iterate through the string
+        for (int32_t i = 0; i < str.length(); ++i) {
+            char c = str.data()[i];
+
+            // Handle sign
+            if (i == 0 && c == '+') {
+                continue; // Skip leading plus sign
+            }
+            if (i == 0 && c == '-') {
+                status = kUnitIdentifierSyntaxError;
+                return result;
+            }
+
+            // Handle digits
+            if (c >= '0' && c <= '9') {
+                uint16_t digit = c - '0';
+                result = result * 10 + digit;
+
+                // Check if the result is within the valid range (0 to int64_t::MAX)
+                if (result > max_int64) {
+                    status = kUnitIdentifierSyntaxError;
+                    return result;
+                }
+                continue;
+            }
+
+            // Handle 'e' or 'E'
+            if (c == 'e' || c == 'E') {
+                exponentIndex = i + 1;
+                break;
+            }
+
+            // Invalid character
+            status = kUnitIdentifierSyntaxError;
+            return result;
+        }
+
+        // Handle exponent
+        for (int i = exponentIndex; i < str.length(); ++i) {
+            char c = str.data()[i];
+
+            // handle sign
+            if (i == exponentIndex && c == '+') {
```

Comment:
Is `+` actually a legal character in a unit identifier?

Reply:
Only if it is part of the unit constant:

For example, "meter-per-+10".

However, it looks a bit weird, and when the user returns the identifier, it will be removed.

This is now handled by the `StringToDoubleConverter`.